comp/core/agenttelemetry/impl/agenttelemetry.go

```diff
@@ -245,6 +251,48 @@ func (a *atel) aggregateMetricTags(mCfg *MetricConfig, mt dto.MetricType, ms []*
 	return maps.Values(amMap)
 }
 
+// Adjust prometheus counter value to be present on each collection. Last
+// counter values need to be cached to calculate the difference. It is a
+// two phase process. First prepares, calculates and collects keys to
+// avoid memory allocations under a lock. Second phase calculates
+// difference between last stored value and cache counter value as a new value
+func (a *atel) adjustPrometheusCounterValue(metricName string, metrics []*dto.Metric) {
+	if len(metrics) > 0 {
+		keyNames := make([]string, 0, len(metrics))
+		for _, m := range metrics {
+			tags := m.GetLabel()
+			if len(tags) == 0 {
+				keyNames = append(keyNames, metricName)
+			} else {
+				// Sort tags to stability of the key
+				sortedTags := cloneLabelsSorted(tags)
+				var builder strings.Builder
+				builder.WriteString(metricName)
+				for _, tag := range sortedTags {
+					builder.WriteString(makeLabelPairKey(tag))
+				}
+				keyNames = append(keyNames, builder.String())
+			}
+		}
+
+		// Adjust the counter values (if found) and cache its current value
+		a.prevCounterValuesMU.Lock()
+		for i, m := range metrics {
+			key := keyNames[i]
+			curValue := m.GetCounter().GetValue()
+
+			// Adjust the counter value if found
+			if prevValue, ok := a.prevCounterValues[key]; ok {
+				*m.GetCounter().Value -= prevValue
+			}
+
+			// Upsert the cache of previous counter values
+			a.prevCounterValues[key] = curValue
+		}
+		defer a.prevCounterValuesMU.Unlock()
```

Comment:
```suggestion
		a.prevCounterValuesMU.Unlock()
```

Reply:
It seems that the `defer` at the last line is equivalent as not having a defer at all 😄